porch/controllers/remoterootsync/pkg/applyset/applyset.go

```diff
@@ -0,0 +1,216 @@
+	// restMapper is used to map object kind to resources, and to know if objects are cluster-scoped.
+	restMapper meta.RESTMapper
+	// patchOptions holds the options used when applying, in particular the fieldManager
+	patchOptions metav1.PatchOptions
+
+	// mutex guards trackers
+	mutex sync.Mutex
+	// trackers is a (mutable) pointer to the (immutable) objectTrackerList, containing a list of objects we are applying.
+	trackers *objectTrackerList
+}
+
+// Options holds the parameters for building an ApplySet.
+type Options struct {
+	// Client is the dynamic kubernetes client used to apply objects to the k8s cluster.
+	Client dynamic.Interface
+	// RESTMapper is used to map object kind to resources, and to know if objects are cluster-scoped.
+	RESTMapper meta.RESTMapper
+	// PatchOptions holds the options used when applying, in particular the fieldManager
+	PatchOptions metav1.PatchOptions
+}
+
+// New constructs a new ApplySet
+func New(options Options) (*ApplySet, error) {
+	a := &ApplySet{
+		client:       options.Client,
+		restMapper:   options.RESTMapper,
+		patchOptions: options.PatchOptions,
+	}
+	a.trackers = &objectTrackerList{}
+	return a, nil
+}
+
+// ReplaceAllObjects is used to replace the desired state of all the objects.
+// Any objects not specified are removed from the "desired" set.
+func (a *ApplySet) ReplaceAllObjects(objects []*unstructured.Unstructured) error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	newTrackers := a.trackers.replaceAllObjects(objects)
+	a.trackers = newTrackers
+
+	return nil
+}
+
+// ApplyResults contains the results of an Apply operation.
+type ApplyResults struct {
+	total             int
+	applySuccessCount int
+	applyFailCount    int
+	healthyCount      int
+	unhealthyCount    int
+}
+
+// AllApplied is true if the desired state has been successfully applied for all objects.
+// Note: you likely also want to check AllHealthy, if you want to be sure the objects are "ready".
+func (r *ApplyResults) AllApplied() bool {
+	r.checkInvariants()
+
+	return r.applyFailCount == 0
+}
+
+// AllHealthy is true if all the objects have been applied and have converged to a "ready" state.
+// Note that this is only meaningful if AllApplied is true.
+func (r *ApplyResults) AllHealthy() bool {
+	r.checkInvariants()
+
+	return r.unhealthyCount == 0
+}
+
+// checkInvariants is an internal function that warns if the object doesn't match the expected invariants.
+func (r *ApplyResults) checkInvariants() {
+	if r.total != (r.applySuccessCount + r.applyFailCount) {
+		klog.Warningf("consistency error (apply counts): %#v", r)
+	} else if r.total != (r.healthyCount + r.unhealthyCount) {
+		// This "invariant" only holds when all objects could be applied
+		klog.Warningf("consistency error (healthy counts): %#v", r)
+	}
+}
+
+// applySkipped records that the apply of an object was skipped because it was already applied.
+func (r *ApplyResults) applySkipped(gvk schema.GroupVersionKind, nn types.NamespacedName) {
```

Comment:
will future PRs utilize the gvk and nn parameters?

Reply:
That's the intent, yes.  We're able to expose per-object status, and often that will be exposed to users, particularly on failure.  We don't do so yet, so right now we're just keeping the count.

I started passing this because pruning gets tricky - we don't always know the version for example, depending on how we are tracking ownership - so I wanted to track what information we actually had.